### PR TITLE
[#11383] Add detection of PRs linking to closed issues

### PR DESCRIPTION
**Pull request by itstrueitstrueitsrealitsreal**
Part of #11383

**Outline of Solution**
Modified `pr.yml` to check for the status of the issue that newly created pull requests are solving using the `pr.data.state.description` attribute from GitHub's [REST API for issues](https://docs.github.com/en/rest/issues/issues?apiVersion=2022-11-28#get-an-issue), allowing users to be notified that the issue they are trying to work on is already closed, as per [Automation Hero](https://github.com/TEAMMATES/teammates/wiki/Automation-Hero) in the wiki. Tested on my personal fork.

### CODE REVIEW

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -13,31 +13,35 @@ jobs:
...
+            }
+            let body = `Hi @${pr.data.user.login}, thank you for your interest in contributing to TEAMMATES!
+              However, your PR does not appear to follow our [contribution guidelines](https://teammates.github.io/teammates/process.html#step-4-submit-a-pr):\n\n`
+            if (!isTitleValid) {
+              body += "- Title must start with the issue number the PR is fixing in square brackets, e.g. `[#<issue-number>]`\n"
+            }
+            if (!isDescriptionValid) {
+              body += "- Description must reference the issue number the PR is fixing, e.g. `Fixes #<issue-number>` (or `Part of #<issue-number>` if the PR does not address the issue fully)\n"
+            }
+            if (!isIssueOpen) {
```
Comment (domoberzin) [inline .github/workflows/pr.yml:38]: Based on this branching, it is possible that the body contains the message for not having a `Fixes #<issue-number>` as well as the message for not addressing an open issue right? I would assume logic should work  something like this:
- If isDescriptionValid is false, it means the PR description doesn't contain a valid reference to an issue number.
- If isIssueOpen is checking whether the referenced issue is open (assuming it was implemented correctly), it wouldn't be able to perform this check if there's no valid issue reference in the description.

Based on the above, a more logical flow should look something like:

- Check if the description is valid (contains a proper issue reference)
- Only if the description is valid, then fetch and check the state of the referenced issue (I believe the issue number can be extracted using capture groups)

To use a capture group, the extraction process might look something like this (this is just an example, do test it out and verify)

```
const descriptionRegex = /(?:[Ff]ix(?:es|ed)?|[Cc]lose(?:s|d)?|[Rr]esolve(?:s|d)?|[Pp]art [Oo]f) #(\d+)/;
const extractIssueNum = (description) => {
  const match = description.match(descriptionRegex);
  return match ? match[1] : null;
};
const issueNumber = extractIssueNumber(pr.data.body);
```

**Review by domoberzin — CHANGES_REQUESTED** [review]
Thanks for working on this, but I think we may need to make some changes

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -13,31 +13,35 @@ jobs:
...
-          })
+      - uses: actions/github-script@v7
+        with:
+          script: |
+            const pr = await github.rest.pulls.get({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              pull_number: context.issue.number
+            })
+            const isIssueOpen = pr.data.state.description === 'open';
```
[the remark this answers is not in the document]
Comment (itstrueitstrueitsrealitsreal) [reply .github/workflows/pr.yml:24]: Thanks for the review @domoberzin, the link you provided was to the response for pull requests, but the workflow file queries for issues, and the schema of the response for issues are:
```json
//...

{
  "title": "Issue",
  "description": "Issues are a great way to keep track of tasks, enhancements, and bugs for your projects.",
  "type": "object",
  "properties": {
    "id": {
      "type": "integer",
      "format": "int64"
    },
    "node_id": {
      "type": "string"
    },
    "url": {
      "description": "URL for the issue",
      "type": "string",
      "format": "uri",
      "examples": [
        "https://api.github.com/repositories/42/issues/1"
      ]
    },
    "repository_url": {
      "type": "string",
      "format": "uri"
    },
    "labels_url": {
      "type": "string"
    },
    "comments_url": {
      "type": "string",
      "format": "uri"
    },
    "events_url": {
      "type": "string",
      "format": "uri"
    },
    "html_url": {
      "type": "string",
      "format": "uri"
    },
    "number": {
      "description": "Number uniquely identifying the issue within its repository",
      "type": "integer",
      "examples": [
        42
      ]
    },
    "state": {
      "description": "State of the issue; either 'open' or 'closed'", // this is the relevant field
      "type": "string",
      "examples": [
        "open"
      ]
    },

//...



```

[the remark this answers is not in the document]
Comment (domoberzin) [reply .github/workflows/pr.yml:24]: @itstrueitstrueitsrealitsreal correct me if im wrong, but the line above
```
 const pr = await github.rest.pulls.get({
              owner: context.repo.owner,
              repo: context.repo.repo,
              pull_number: context.issue.number
            })
 ```
 
 is querying for PRs, not issues right?

[the remark this answers is not in the document]
Comment (itstrueitstrueitsrealitsreal) [reply .github/workflows/pr.yml:24]: Alright, I seem to have made a mistake w.r.t. to pulls and issues, will rectify this, thanks!

**Review by domoberzin — APPROVED** [review]
LGTM

**Review by mingyuanc — APPROVED** [review]
LGTM, interesting solution!

**Pull request merged**